drivers/usb/udc/udc_numaker.c

```diff
@@ -635,61 +1064,183 @@ static void numaker_usbd_ep_copy_to_user(struct numaker_usbd_ep *ep_cur, uint8_t
 	} else {
 		data_rmn = (ep_base->MXPLD & USBD_MXPLD_MXPLD_Msk) >> USBD_MXPLD_MXPLD_Pos;
 	}
+#elif DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_hsusbd)
+	if (ep_cur->ep_hw_idx == CEP) {
+		data_rmn = (base->CEPDATCNT & HSUSBD_CEPDATCNT_DATCNT_Msk) >>
+			   HSUSBD_CEPDATCNT_DATCNT_Pos;
+	} else {
+		data_rmn = (ep_base->EPDATCNT & HSUSBD_EPDATCNT_DATCNT_Msk) >>
+			   HSUSBD_EPDATCNT_DATCNT_Pos;
+	}
+#endif
 
 	*size_p = MIN(*size_p, data_rmn);
 
+#if DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_usbd)
 	bytecpy(usrbuf, (uint8_t *)dmabuf_addr, *size_p);
+#elif DT_HAS_COMPAT_STATUS_OKAY(nuvoton_numaker_hsusbd)
+	if (ep_cur->ep_hw_idx == CEP) {
+		uint8_t *usrbuf_pos = usrbuf;
+		uint32_t rmn = *size_p;
+
+		while (rmn && !(base->CEPINTSTS & HSUSBD_CEPINTSTS_BUFEMPTYIF_Msk)) {
+			*usrbuf_pos++ = base->CEPDAT_BYTE;
+			rmn--;
+		}
+
+		*size_p -= rmn;
+	} else {
+#if defined(CONFIG_UDC_NUMAKER_DMA)
+		int err = numaker_xusbd_ep_xfer_user_dma(ep_cur, usrbuf, size_p);
+
+		if (err < 0) {
+			return err;
+		}
+#else
+		uint8_t *usrbuf_pos = usrbuf;
+		uint32_t rmn = *size_p;
+
+		while (rmn && !(ep_base->EPINTSTS & HSUSBD_EPINTSTS_BUFEMPTYIF_Msk)) {
+			*usrbuf_pos++ = ep_base->EPDAT_BYTE;
+			rmn--;
+		}
+
+		*size_p -= rmn;
+#endif
+	}
+#endif
+
```

Comment:
This part and similar ones below do not look good. Please rework it/move it to specific functions.

Reply:
They are moved to separate functions for clear.